packages/core/src/lib/passkey/verify.ts

```diff
@@ -0,0 +1,249 @@
+  WebAuthnVerificationError,
+} from "../../errors"
+import { decodeSignedCookie } from "../cookie"
+
+export type UserData = {
+  user: AdapterUser
+  account: Account
+  authenticator?: AdapterAuthenticator
+}
+
+/**
+ * Verify a passkey registration response.
+ * It checks the challenge cookie, verifies the response, and updates the user's authenticators.
+ *
+ * @param options
+ * @param reqCookies request cookies
+ * @param response response object from the client's `startAuthentication`
+ * @returns Whether the authentication was successful.
+ */
+export async function verifyAuthentication(
+  options: Options,
+  reqCookies: RequestInternal["cookies"],
+  response: unknown
+): Promise<UserData | string> {
+  const { adapter: _adapter, provider, logger } = options
+  const adapter: Adapter | undefined = _adapter
+
+  // Validate that the adapter is defined and implements the required methods
+  if (!adapter)
+    throw new MissingAdapter(
+      "WebaAuthn verifyAuthentication requires an adapter."
+    )
+  assertAdapterImplementsMethods(
+    "WebaAuthn verifyAuthentication requires an adapter that implements",
+    adapter,
+    ["getAuthenticator", "getUserByAccount", "updateAuthenticatorCounter"]
+  )
+
+  // Basic response type check
+  if (
+    typeof response !== "object" ||
+    response === null ||
+    !("id" in response) ||
+    typeof response.id !== "string"
+  ) {
+    return "Invalid response."
+  }
+
+  // Get cookie with challenge
+  const [cookie] = await decodeSignedCookie<PasskeyOptionsCookieData>(
+    "challenge",
+    reqCookies,
+    options
+  )
+  if (!cookie) {
+    return "Missing challenge cookie."
+  }
+  const { challenge: expectedChallenge } = cookie
+
+  // Get the authenticator from the database
+  const uintID = new Uint8Array(Buffer.from(response.id, "base64"))
+  const authenticator = (await adapter.getAuthenticator(uintID)) ?? null
+  if (!authenticator) {
+    logger.debug(`Authenticator not found.`, { id: response.id })
+    return `Authenticator not found.`
+  }
+
+  // Do the actual verification
+  let verification: VerifiedAuthenticationResponse
+  try {
+    verification = await verifyAuthenticationResponse({
+      response: response as AuthenticationResponseJSON,
+      authenticator,
+      expectedChallenge,
+      expectedOrigin: provider.relayingParty.origin,
+      expectedRPID: provider.relayingParty.id,
+      requireUserVerification: true,
+    })
+  } catch (e: any) {
+    const err = new WebAuthnVerificationError(e)
+    logger.error(err)
+
+    return err.message
+  }
+
+  const { verified, authenticationInfo } = verification
+
+  // make sure the response is verified
+  if (!verified) {
+    return "Failed to verify response."
+  }
+
+  // Update the authenticator counter
+  try {
+    const newCounter = authenticationInfo.newCounter
+    await adapter.updateAuthenticatorCounter(authenticator, newCounter)
+  } catch (e: any) {
+    logger.error(new AdapterError(e))
```

Comment:
Should we tolerate this counter update failure like this, or we abort and return an error?

Reply:
We shouldn't. Failure to update the counter will make the authenticator unusable.
Fixed in 4e88ec13e45eeb27d5f21a2e750fc9d02390f2e4

btw, if this does happen (failure to update counter), the dev will probably need to go into the database and manually update using the logged `newCounter` value.